# Data Cleaning with R 

### We begin by determining the best way to import the file (s). Upon opening the actual txt file, we discover the file is delimited (seperated by commas and tabs).

Import relevant libraries

In [247]:
library(tidyverse)
library(data.table)
library(stringr)
library(readr)
library(plyr)
library(tidyr)
library(dplyr)

In [255]:
# reading in the data 
CO2_ls = list.files(path = '../R/Data/Surface CO2 from NOAA/')
CO2_ls

[1] "co2_alt_surface-flask_1_ccgg_month.txt"
[2] "co2_bhd_surface-flask_1_ccgg_month.txt"
[3] "co2_crz_surface-flask_1_ccgg_month.txt"
[4] "co2_eic_surface-flask_1_ccgg_month.txt"
[5] "co2_lln_surface-flask_1_ccgg_month.txt"
[6] "co2_nmb_surface-flask_1_ccgg_month.txt"

In [305]:
path <-  '../R/Data/Surface CO2 from NOAA/'
file_name <- CO2_ls[1]
df <- read.delim(paste0(path,file_name), skip = 0, header = FALSE, as.is=TRUE, dec=".", sep = ",", na.strings=c(" ", "",'NA'), strip.white = TRUE)
head(df, n=10)

V1                                                                             
1  # number_of_header_lines: 69                                                   
2  # comment:                                                                     
3  # comment:  ************ USE OF ESRL GMD DATA ****************                 
4  # comment:                                                                     
5  # comment: These data are made freely available to the public and the          
6  # comment: scientific community in the belief that their wide dissemination    
7  # comment: will lead to greater understanding and new scientific insights.     
8  # comment: The availability of these data does not constitute publication      
9  # comment: of the data.  NOAA relies on the ethics and integrity of the user to
10 # comment: ensure that ESRL receives fair credit for their work.  If the data

In [306]:
(df[1,1])
(num_header_lines<- sub(".*: ", "", df[1,1]) %>% as.numeric())

[1] "# number_of_header_lines: 69"

[1] 69

In [309]:
df <- read.delim(paste0(path, file_name), skip = (num_header_lines-1), header = FALSE, as.is=TRUE, dec=".", sep = "", na.strings=c(" ", "",'NA'), strip.white = TRUE)
head(df, n=13)

V1  V2           V3   V4     V5    V6   
1  #   data_fields: site year   month value
2  ALT 1985         6    349.94 NA    NA   
3  ALT 1985         7    343.98 NA    NA   
4  ALT 1985         8    337.96 NA    NA   
5  ALT 1985         9    339.09 NA    NA   
6  ALT 1985         10   342.85 NA    NA   
7  ALT 1985         11   345.66 NA    NA   
8  ALT 1985         12   348.97 NA    NA   
9  ALT 1986         1    351.34 NA    NA   
10 ALT 1986         2    352.13 NA    NA   
11 ALT 1986         3    352.74 NA    NA   
12 ALT 1986         4    353.12 NA    NA   
13 ALT 1986         5    353.32 NA    NA

In [310]:
(hdr <- c(df[1,4], df[1,5], df[1,6]))
head(df <- df %>% select(2:4) %>% slice(2:n()))

# df <- df %>% select(2:4)
# head(df <- slice(df, 2:n()))
is.na(hdr)

names(df) <- hdr
head(df)

[1] "year"  "month" "value"

V2   V3 V4    
1 1985 6  349.94
2 1985 7  343.98
3 1985 8  337.96
4 1985 9  339.09
5 1985 10 342.85
6 1985 11 345.66

[1] FALSE FALSE FALSE

year month value 
1 1985 6     349.94
2 1985 7     343.98
3 1985 8     337.96
4 1985 9     339.09
5 1985 10    342.85
6 1985 11    345.66

In [311]:
df %>% summarise_all(funs(sum(is.na(.))))

year month value
1 0    0     0

In [313]:
(class(df[1,3]))
df %>% select(2) %>% unique

[1] "character"

month
1  6    
2  7    
3  8    
4  9    
5  10   
6  11   
7  12   
8  1    
9  2    
10 3    
11 4    
12 5

In [314]:
df$month <- plyr::mapvalues(df$month, 
          from=c("12","11","10",'9', '8', '7', '6', '5', '4','3', '2','1'), 
          to=c("Dec","Nov","Oct", 'Sep', 'Aug', 'Jul', 'Jun', 'May', 'Apr', 'Mar', 'Feb', 'Jan'))

In [316]:
df %>% head()
df %>% select(2) %>% unique

year month value 
1 1985 Jun   349.94
2 1985 Jul   343.98
3 1985 Aug   337.96
4 1985 Sep   339.09
5 1985 Oct   342.85
6 1985 Nov   345.66

month
1  Jun  
2  Jul  
3  Aug  
4  Sep  
5  Oct  
6  Nov  
7  Dec  
8  Jan  
9  Feb  
10 Mar  
11 Apr  
12 May

In [298]:
df_copy <- data.frame(df)
(tt <- data.frame())

<0 x 0 matrix>

In [296]:
(rownames(tt) <- df_copy$year %>% as.numeric()%>%unique())

ERROR: Error in `.rowNamesDF<-`(x, value = value): invalid 'row.names' length


In [289]:
# (almost_there <- t(df_copy) %>% as_data_frame())

# (names(almost_there) <-df_copy[,3])

# (closer <- almost_there %>% slice(-3))

V1     V2     V3     V4     V5     V6     V7     V8     V9     V10    ...
1 ALT    ALT    ALT    ALT    ALT    ALT    ALT    ALT    ALT    ALT    ...
2 1985   1985   1985   1985   1985   1985   1985   1986   1986   1986   ...
3 Jun    Jul    Aug    Sep    Oct    Nov    Dec    Jan    Feb    Mar    ...
4 349.94 343.98 337.96 339.09 342.85 345.66 348.97 351.34 352.13 352.74 ...
  V377   V378   V379   V380   V381   V382   V383   V384   V385   V386  
1 ALT    ALT    ALT    ALT    ALT    ALT    ALT    ALT    ALT    ALT   
2 2017   2017   2017   2017   2017   2017   2017   2017   2017   2017  
3 Mar    Apr    May    Jun    Jul    Aug    Sep    Oct    Nov    Dec   
4 413.38 413.62 413.26 410.16 402.56 396.90 397.69 402.83 407.79 410.79

[1] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
 [13] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
 [25] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
 [37] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
 [49] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
 [61] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
 [73] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
 [85] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
 [97] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
[109] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
[121] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
[133] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
[145] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
[157] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
[169] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
[181] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
[193] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
[205] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
[217] "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May"
[229] "Jun" "Jul" "Aug" "Sep" "Oct" "Apr" "May" "Jun" "Jul" "Aug" "Sep" "Oct"
[241] "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May" "Jun" "Jul" "Aug" "Sep" "Oct"
[253] "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May" "Jun" "Jul" "Aug" "Sep" "Oct"
[265] "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May" "Jun" "Jul" "Aug" "Sep" "Oct"
[277] "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May" "Jun" "Jul" "Aug" "Sep" "Oct"
[289] "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May" "Jun" "Jul" "Aug" "Sep" "Oct"
[301] "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May" "Jun" "Jul" "Aug" "Sep" "Oct"
[313] "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May" "Jun" "Jul" "Aug" "Sep" "Oct"
[325] "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May" "Jun" "Jul" "Aug" "Sep" "Oct"
[337] "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May" "Jun" "Jul" "Aug" "Sep" "Oct"
[349] "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May" "Jun" "Jul" "Aug" "Sep" "Oct"
[361] "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May" "Jun" "Jul" "Aug" "Sep" "Oct"
[373] "Nov" "Dec" "Jan" "Feb" "Mar" "Apr" "May" "Jun" "Jul" "Aug" "Sep" "Oct"
[385] "Nov" "Dec"

Jun    Jul    Aug    Sep    Oct    Nov    Dec    Jan    Feb    Mar    ...
1 ALT    ALT    ALT    ALT    ALT    ALT    ALT    ALT    ALT    ALT    ...
2 1985   1985   1985   1985   1985   1985   1985   1986   1986   1986   ...
3 349.94 343.98 337.96 339.09 342.85 345.66 348.97 351.34 352.13 352.74 ...
  Mar    Apr    May    Jun    Jul    Aug    Sep    Oct    Nov    Dec   
1 ALT    ALT    ALT    ALT    ALT    ALT    ALT    ALT    ALT    ALT   
2 2017   2017   2017   2017   2017   2017   2017   2017   2017   2017  
3 413.38 413.62 413.26 410.16 402.56 396.90 397.69 402.83 407.79 410.79

## Saving files under new names and new and directories
Extracting the station number, city and province for file name use. 

In [31]:
df = read.delim('R/Data/Homog_monthly_max_temp/mx230N002.txt', skip = 0, header = FALSE, as.is=TRUE, dec=".", sep = ",", na.strings=c(" ", "",'NA'), strip.white = TRUE)

(station_number <- select(df, V1)[1,1])
(city <- select(df, V2)[1,1])
(province <- select(df, V3)[1,1])


[1] "230N002"

[1] "LUPIN"

[1] "NU"

In [32]:
stationNum_city_prov <- paste(select(df, V1)[1,1], trimws(select(df, V2)[1,1]), province <- select(df, V3)[1,1], sep='_')
stationNum_city_prov

[1] "230N002_LUPIN_NU"

## Result 

338 text files were cleaned, per directory
* Directories: Homog_monthly_min_temp, Homog_monthly_max_temp, Homog_monthly_mean_temp

As shown above, new text files and directories are created
* New Directories: Homog_monthly_min_temp_cleaned, Homog_monthly_max_temp_cleaned, Homog_monthly_mean_temp_cleaned

 

## Intended use 
All data is cleaned prior to the shinyApp being used. The cleaned data is reloaded and processed into a single data frame on app load. 


### Input data frame. 
Depending on the month and year the user selects, the data is trimmed before any statistical method.


![input df](input_df.PNG)

### Output data frame. 
After the desired statistical analysis, we shift our focus to different variables and a new data frame.


![output df](output_df.PNG)

## In conclusion, the data does not appear to have many challenges. The ease and use of the tidy verse package, is one that can be consistently used. Helping many, understand the data they are dealing with.